content/news/20.4networking-rework.md

```diff
@@ -0,0 +1,231 @@
+
+### Payload discrimination
+When considering this system, you might, right-fully so, ask how the system keeps the different payload types apart from each other. This is achieved by the system it-self. It writes a discriminator ID to the connection before invoking the actually writing mechanic for a payload. Conversely on the client side, the discriminator is read first, before a reader is looked up so that the rest of the payload can be read.
+
+{{< notice note >}}
+The system handles reading and writing the discriminator it-self, as a modder you do not need to take care of this yourself.
+{{< /notice >}}
+
+The value of the discriminator during writing is retrieved from the `CustomPacketPayload#id()` method, and is not allowed to be null.
+The value against which the id, read from the connection, is compared to find a reader, is the one that is given to the registrar as the first argument.
+It is as such of the upmost importance that you return the same value, as in the same resource location, from both the `id()` method on the payload instance, as well as given the same value to the registrar when registering your payload.
+
+{{< notice tip >}}
+We recommend that you store your ID in a `public static final ResourceLocation ID = new ResourceLocation("mod_id", "payload_id")` field, and reference that in both places.
+{{< /notice >}}
+
+{{< notice warning >}}
+Given that the id is used as a discriminator, it is important that you use a unique value, especially for the ResourceLocations path section for each payload type. If you try to register the same id twice, the registrar will throw an exception. If you try to register an id with a namespace other than the one the registrar is for, the registrar will throw an exception. You are free to request registrars for other namespaces then your own.
+{{< /notice >}}
+
+### Payload reading
+Payload reading happens via the vanilla method, implementing the `FriendlyByteBuf.Reader<T>` functional interface. During the registration of the payload type, as can be seen above, you need to pass an implementation of this interface so that a new instance of the payload can be created by the system when a custom payload packet with it as payload arrives at the receiving end.
+
+{{< notice tip >}}
+As we recommend that your payload implementations are `record`s in java, instead of classes, we also recommend that you create a custom constructor with the record, to read the records fields from the buffer. This constructor can then be passed as a method reference for the reader implementation. So if `SimplePayload(String somethign) {}` is your normal record, then adding `SimplePayload(FriendlyByteBuf buf) { this(buf.readUtf()); }` as a constructor to the `SimplePayload` record will allow you to pass it as a method reference `SimplePayload::new` to the registrar when it asks for an implementation of `FriendlyByteBuf.Reader<SimplePayload>`
```

Comment:
Since you are supposed to create immutable data carriers (hence records), I see no problem here. An advanced modder that needs this can create a second constructor that further processes your values or create a custom reader.